ui/remote_button.js

```diff
@@ -0,0 +1,201 @@
+        'shaka-current-selection-span');
+    label.appendChild(this.remoteCurrentSelectionSpan_);
+    this.remoteButton_.appendChild(label);
+    this.parent.appendChild(this.remoteButton_);
+
+    /** @private {number} */
+    this.callbackId_ = -1;
+
+    // Setup strings in the correct language
+    this.updateLocalizedStrings_();
+
+    shaka.ui.Utils.setDisplay(this.remoteButton_, false);
+
+    if (!this.video.remote || this.video.disableRemotePlayback) {
+      this.remoteButton_.classList.add('shaka-hidden');
+    } else {
+      this.eventManager.listen(
+          this.localization, shaka.ui.Localization.LOCALE_UPDATED, () => {
+            this.updateLocalizedStrings_();
+          });
+
+      this.eventManager.listen(
+          this.localization, shaka.ui.Localization.LOCALE_CHANGED, () => {
+            this.updateLocalizedStrings_();
+          });
+
+      this.eventManager.listen(this.controls, 'caststatuschanged', () => {
+        this.updateRemoteState_();
+      });
+
+      this.eventManager.listen(this.remoteButton_, 'click', () => {
+        this.video.remote.prompt();
+      });
+
+      this.eventManager.listen(this.video.remote, 'connect', () => {
+        this.updateRemoteState_();
+      });
+
+      this.eventManager.listen(this.video.remote, 'connecting', () => {
+        this.updateRemoteState_();
+      });
+
+      this.eventManager.listen(this.video.remote, 'disconnect', () => {
+        this.updateRemoteState_();
+      });
+
+      this.eventManager.listen(this.player, 'loaded', () => {
+        this.updateRemoteState_();
+      });
+
+      this.updateRemoteState_();
+    }
+  }
+
+  /** @override */
+  release() {
+    if (this.video.remote && this.callbackId_ != -1) {
+      this.video.remote.cancelWatchAvailability(this.callbackId_);
+    }
+
+    super.release();
+  }
+
+  /**
+   * @private
+   */
+  async updateRemoteState_() {
+    if (this.controls.getCastProxy().canCast() &&
+        this.controls.isCastAllowed()) {
+      shaka.ui.Utils.setDisplay(this.remoteButton_, false);
+      if (this.callbackId_ != -1) {
+        this.video.remote.cancelWatchAvailability(this.callbackId_);
+        this.callbackId_ = -1;
+      }
+    } else if (this.video.remote.state == 'disconnected') {
+      const handleAvailabilityChange = (availability) => {
+        if (this.player) {
+          const loadMode = this.player.getLoadMode();
+          const srcMode = loadMode == shaka.Player.LoadMode.SRC_EQUALS;
+          shaka.ui.Utils.setDisplay(
+              this.remoteButton_, srcMode && availability);
+        } else {
+          shaka.ui.Utils.setDisplay(this.remoteButton_, false);
+        }
+      };
+      try {
+        if (this.callbackId_ != -1) {
+          await this.video.remote.cancelWatchAvailability(this.callbackId_);
+          this.callbackId_ = -1;
+        }
+      } catch (e) {
+        // Ignore this error.
+      }
+      try {
+        const id = await this.video.remote.watchAvailability(
```

Comment:
Does this account for MediaSource playbacks somehow?

Reply:
In the method  handleAvailabilityChange